File: Java/IntelligentInSitesAPI/src/com/intelligentinsites/api/APIParams.java
package com.intelligentinsites.api;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.AbstractMap.SimpleEntry;

import org.apache.http.NameValuePair;
import org.apache.http.message.BasicNameValuePair;

public class APIParams {
	private Map<String, List<Object>> params;
	
	public APIParams() {
		params = new HashMap<String, List<Object>>();
	}
	
	/**
	 * Adds a File parameter to the collection. This would typically be used in an HTTP multipart/form-data POST request.
	 * 
	 * @param name	the parameter name
	 * @param value	the File to add as a parameter
	 */
	public void add(String name, File value) {
		addObject(name, value);
	}
	
	/**
	 * Adds a String parameter to the collection.
	 * 
	 * @param name	the parameter name
	 * @param value	the value of the parameter
	 */
	public void add(String name, String value) {
		addObject(name, value);
	}
	
	/**
	 * Removes all parameters from the collection.
	 */
	public void clear() {
		params.clear();
	}
	
	/**
	 * Returns the names of all parameters in the collection.
	 * 
	 * @return	a {@link Set} of all parameter names in the collection
	 */
	public Set<String> getNames() {
		return params.keySet();
	}
	
	/**
	 * Returns the value of a parameter in the collection.
	 * If the parameter has multiple values, this returns a {@link List} of those values.
	 * 
	 * @param name	the name of a parameter
	 * @return		the value of the given parameter. If the parameter has multiple values, this returns a {@link List} of those values.
	 */
	public Object getValue(String name) {
		List<Object> value = params.get(name);
		if (value != null) {
			if (value.size() == 1) {
				return value.get(0);
			}
		}
		return value;
	}
	
	/**
	 * Returns all the parameters whose values are Strings.
	 * 
	 * @return	a List of {@link org.apache.http.NameValuePair} objects representing all the String parameters defined in the collection
	 */
	public List<NameValuePair> getStringParams() {
		ArrayList<NameValuePair> stringParams = new ArrayList<NameValuePair>();
        for (Map.Entry<String, List<Object>> param : params.entrySet()) {
            List<Object> value = param.getValue();
            for (Object o : value) {
            	if (o instanceof String) {
            		stringParams.add(new BasicNameValuePair(param.getKey(), o.toString()));
            	}
            }
        }
        return stringParams;
	}
	
	/**
	 * Returns all the parameters whose values are Files.
	 * 
	 * @return	a List of {@link java.util.Map.Entry} objects representing all the File parameters defined in the collection
	 */
	public List<Entry<String, File>> getFileParams() {
		ArrayList<Entry<String, File>> fileParams = new ArrayList<Entry<String, File>>();
        for (Map.Entry<String, List<Object>> param : params.entrySet()) {
            List<Object> value = param.getValue();
            for (Object o : value) {
            	if (o instanceof File) {
            		fileParams.add(new SimpleEntry<String, File>(param.getKey(), (File)o));
            	}
            }
        }
        return fileParams;
	}
	
	/**
	 * Returns whether the given parameter name has been assigned a value in this instance.
	 * 
	 * @param name	the name of a parameter
	 * @return		true if the given parameter name has been assigned a value
	 */
	public boolean isParameterSet(String name) {
		return params.containsKey(name);
	}
	
	private void addObject(String name, Object value) {
		if (params.containsKey(name)) {
			params.get(name).add(value);
		} else {
			ArrayList<Object> valueList = new ArrayList<Object>();
			valueList.add(value);
			params.put(name, valueList);
		}
	}
}

File: DotNet/IntelligentInSites.Api/IntelligentInSites.Api/APIParams.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using System.Collections;

namespace IntelligentInSites.Api.Rest
{
    public class APIParams : IEnumerable
    {
        private Dictionary<String, List<Object>> Params;

        public APIParams()
        {
            this.Params = new Dictionary<String, List<Object>>();
        }

        /// <summary>
        /// Adds a String parameter to the collection.
        /// </summary>
        /// <param name="name">the parameter name</param>
        /// <param name="value">the value of the parameter</param>
        public void Add(string name, string value)
        {
            AddObject(name, value);
        }

        /// <summary>
        /// Adds a byte[] parameter to the collection. This would typically be used in an HTTP multipart/form-data POST request.
        /// </summary>
        /// <param name="name">the parameter name</param>
        /// <param name="value">the value of the parameter</param>
        public void Add(string name, byte[] value)
        {
            AddObject(name, value);
        }

        /// <summary>
        /// Adds an int parameter to the collection.
        /// </summary>
        /// <param name="name">the parameter name</param>
        /// <param name="value">the value of the parameter</param>
        public void Add(string name, int value)
        {
            AddObject(name, value);
        }

        /// <summary>
        /// Adds a bool parameter to the collection.
        /// </summary>
        /// <param name="name">the parameter name</param>
        /// <param name="value">the value of the parameter</param>
        public void Add(string name, bool value)
        {
            AddObject(name, value);
        }

        /// <summary>
        /// Adds a decimal parameter to the collection.
        /// </summary>
        /// <param name="name">the parameter name</param>
        /// <param name="value">the value of the parameter</param>
        public void Add(string name, decimal value)
        {
            AddObject(name, value);
        }

        /// <summary>
        /// Removes all parameters from the collection.
        /// </summary>
        public void Clear()
        {
            Params.Clear();
        }

        public IEnumerator GetEnumerator()
        {
            foreach (KeyValuePair<String, List<object>> pair in Params)
            {
                foreach (object value in pair.Value)
                {
                    if (value is byte[])
                    {
                        yield return new KeyValuePair<string, object>(pair.Key, (byte[])value);
                    }
                    else
                    {
                        yield return new KeyValuePair<string, object>(pair.Key, value.ToString());
                    }
                }
            }
        }

        
        /// <summary>
        /// Returns the names of all parameters in the collection.
        /// </summary>
        /// <returns>A KeyCollection containing all parameter names in the collection.</returns>
        public Dictionary<string, List<object>>.KeyCollection GetNames()
        {
            return Params.Keys;
        }

        /// <summary>
        /// Returns the value of a parameter in the collection.
        /// If the parameter has multiple values, this returns a List of those values.
        /// </summary>
        /// <param name="name">The name of a parameter.</param>
        /// <returns>The value of the given parameter. If the parameter has multiple values, this returns a List of those values.</returns>
        public object GetValue(string name)
        {
            List<object> value = Params[name];
            if (value != null)
            {
                if (value.Count == 1)
                {
                    return value[0];
                }
            }
            return value;
        }

        /// <summary>
        /// Returns whether the given parameter name has been assigned a value in this instance.
        /// </summary>
        /// <param name="name">The name of a parameter.</param>
        /// <returns>True if the given parameter name has been assigned a value.</returns>
        public bool IsParameterSet(string name)
        {
            return Params.ContainsKey(name);
        }

        private void AddObject(string name, object value)
        {
            if (Params.ContainsKey(name))
            {
                Params[name].Add(value);
            }
            else
            {
                List<object> valueList = new List<object>();
                valueList.Add(value);
                Params.Add(name, valueList);
            }
        }
    }
}
